### ISSUE TITLE
Edit the position of files in a directory structure

### ISSUE THREAD

**Issue by TaniaSchlatter** [IQSS/dataverse#5565]
As part of adding the ability to display file hierarchy at the dataset level (#2249), we want to allow files to be added a directory structure. Completing this work will mean that a user will be able to add a file to a dataset that has files imported from a .zip file or any file, and edit the relative position of files in a dataset.

Anticipated workflow is adding file(s), selecting files from the files table, selecting edit, and editing path(s). 

UI impact is exposing the path, and making it editable.

Changes will be visible in a revised view of the directory structure.

**Comment by djbrooke** [comment]
@mheppler can you add the mockup here? Thank you!

**Comment by mheppler** [comment]
Previous mockups were in Balsamiq, but I whipped up this mock File Upload pg screenshot quick for this issue.

![screen shot 2019-02-26 at 1 28 09 pm](https://user-images.githubusercontent.com/687227/53436989-79bb3600-39ca-11e9-9172-3f33a963dc2a.png)

**Comment by djbrooke** [comment]
- we should allow this editing through the API as well
- we should have some discussion about whether or not to have leading /
- changes should be recorded as a different version
- we should validate the inputs 
- changes here would override changes that were previously made or that we set by uploading a zip

**Comment by scolapasta** [comment]
In the context of this issue, we discussed the current zip upload process of unzipping and whether it should add a folder to the path for the zip file. We decided we would leave it as is, i.e. no folder gets added. If you upload a zip with three files at the top level and a second zip with three files at the top level, the resulting dataset will have six files at the top level. 

Related: #5576

**Comment by sekmiller** [comment]
Per discussion we will track editing all file metadata via api in another issue.

**Comment by mheppler** [comment]
API for file metadata editing is at _Add an API endpoint for File Level metadata #4616_.

**Comment by mheppler** [comment]
Merged `develop` into the branch 6f87904bef11be2d6201115a84b8051fa8f8e639 in order to take advantage all those glorious, new tooltip icons. :dataverseman:

**Comment by mheppler** [comment]
Initial UI clean up of the hierarchy metadata on the dataset, file and edit file pgs.

**EDIT FILES**
<img width="947" alt="screen shot 2019-03-07 at 12 53 10 pm" src="https://user-images.githubusercontent.com/687227/53978067-55e9a580-40d8-11e9-9181-6054940c4ab2.png">

**DATASET FILES TABLE**
<img width="942" alt="screen shot 2019-03-07 at 12 53 35 pm" src="https://user-images.githubusercontent.com/687227/53978075-5bdf8680-40d8-11e9-861f-93d09d9e5566.png">

**FILE METADATA PANEL**
<img width="945" alt="screen shot 2019-03-07 at 12 53 49 pm" src="https://user-images.githubusercontent.com/687227/53978099-6ac63900-40d8-11e9-9d0a-fde28f4607e1.png">

**Comment by mheppler** [comment]
Starting an outline of outstanding to-do issues

- Validation
- Documentation

**Comment by mheppler** [comment]
Updated the Compressed Files section, and added the File Path section to the Dataset + File Management pg of the User Guide. This is very general information, and does not go into any detail about the validation requirements or anything like that.

There was also a reference to the `DirectoryLabel` attribute in the Data Access API pg of the API Guide, but it didn't look any further detail was required in that section.

@scolapasta @sekmiller @dlmurphy -- please take a moment to review these changes, and have at editing whatever you believe is inaccurate or missing.

**Comment by dlmurphy** [comment]
I'll review the guides when I'm back from lunch!

**Comment by pdurbin** [comment]
As mentioned at standup, this bug was found in develop and the latest release (4.11) was found during development: Edit File Metadata does not validate (saves with old (valid) value) #5654

**Comment by landreev** [comment]
I added 1 request/question in the PR.

**Comment by landreev** [comment]
(Aside from that - everything looks good. The thing that I asked about - multiple slashes in the middle - if that was discussed and the final decision was to allow it like that, then please move it directly into QA)

**Comment by landreev** [comment]
That was better, but I'm afraid it still wasn't doing everything we wanted to do. 
From my conversation with @scolapasta, the plan was to replace any backslashes (and any combinations of back and forward slashes) with a single forward slash. 

The new sanitize method was leaving names like "folder1/\/folder2" intact. 

There's also something weird with the validator method. 
It actually fails on any backslashes, unless followed by a forward slash; even though our validation message says back slashes are allowed. (So, "folder1\folder2" - fails; "folder1\/\/folder2" validates and gets saved). 

But I think I solved  it all that with this regex in the sanitize method - because I think that's what we wanted to do all along: 

value = value.replaceAll("[\\\\/][\\\\/]*", "/");

(so the 4 backslashes in a java regex means one backslash. yes)


![Screen Shot 2019-03-21 at 10 26 41 AM](https://user-images.githubusercontent.com/737558/54760689-f3f46a00-4bc6-11e9-9db7-6ad48e1e5e1c.png)

**Comment by landreev** [comment]
Let me quickly check if any of the test methods need to be updated to account for this last change. 
Then I'll move it to QA, unless there are any objections.

**Comment by landreev** [comment]
BTW, why is the sanitize method called in the validator method? 
It's probably redundant - because it's always called in setDirectoryLabel()... but if it weren't - I think it would result in validating the sanitized version, without actually saving it... no? (I may be missing something...)

**Comment by landreev** [comment]
@scolapasta 
Hmm, I'm still looking at that validation regex... it was doing other counter-intuitive things, that I cleaned up. 
two more questions:
we have 49K non-empty directorylabels in prod. already. 
635 of them do not pass the intended validation rule - because they contain "&", "@", "=", "(", ")" or " " (white space). 
So,
1) of these, I think we should allow the whitespace. it's a legit/standard folder character on all OSes by now. 
2) for the ones we don't allow - we'll need to tell other installations to sanitize their existing labels; because it's very likely they have some invalid ones too. how do we do that? I would put together a shell script for our own prod.; and add it to the release notes. But can you think of a better way?

**Comment by scolapasta** [comment]
@landreev I'm ok with both those points. Yes, for any new / stricter validation rules, we'll need to include release notes (and a script like you suggested would also be helpful).

**Comment by landreev** [comment]
Idk, maybe a pure sql script, that we could make flyway execute, is actually an easier solution than a shell script. 
One way or another, it feels like this migration solution needs to be created before we can put this into QA, yes?

**Comment by landreev** [comment]
OK, I'm working on it.

**Comment by landreev** [comment]
@scolapasta 
It's very easy actually, to just  make Flyway do this sanitizing automatically:
d995c39f5

**Comment by kcondon** [comment]
Issues found:
1. Validation text indicates \ is allowed but when used as a folder separator character, throws validation error and does not save.

**Comment by landreev** [comment]
Zip uploads: folder names are now preemptively sanitized; all the invalid characters are removed. I discussed this approach with @scolapasta earlier today and we are going to review and finalize it on Monday.

**Comment by pdurbin** [comment]
A couple of updates from standup:

- Updating the directory structure is now possible from the native API "add" and "replace" endpoint because we merged pull request #5678 into the branch under test for this issue. An update to the API Guide was included.
- I haven't been following very closely the naming convention we are converging on for Flyway scripts but I pushed b2f04e40a to the branch to attempt to document it. This could use some review.
